### ISSUE TITLE
Getting consumer groups details fails with NPE

### ISSUE THREAD

**Issue by aeftef** [Azure/azure-event-hubs-for-kafka#125]
Description
===========

Given a event hubs "topic", using its kafka integration, and an apache spark (2.4.7) structured streaming client reading from the topic, if you try to get consumer groups details it will fail with a NullPointerException (see error here https://github.com/aeftef/azure_eventhubs_kafka_issue/blob/master/consumerGroupsError.txt).

According to event hub kafka integration documentation, Apache Spark streaming (2.4) is a supported configuration, so we guess this should work.

An example of the script used to get consumer group details is here: https://github.com/aeftef/azure_eventhubs_kafka_issue/blob/master/listConsumerGroups.sh (and a complete project to reproduce the case)

Apart from preventing us to query consumer group details, we cannot use some of our prefered kafka ui tools that also use this very same api, I am thinking about kafdrop... So this is quite deal breaker for us to consider using event hubs over native kafka...

How to reproduce
================
1) Start a spark structured streaming client, that reads a event hubs topic.
2) While its running try to get the consumer groups (kafka consumer groups) details. For instance using the kafka-consumer-groups.sh script included with standard kafka.
3) Kafka admin client API will throw a NPE.

Scripts and examples uploaded here: https://github.com/aeftef/azure_eventhubs_kafka_issue

Has it worked previously?
=========================
No

Checklist
=========

**IMPORTANT**: We will close issues where the checklist has not been completed or where adequate information has not been provided.

Please provide the relevant information for the following items:
 - [x] SDK (include version info): `apache spark 2.4.7`
 - [ ] Sample you're having trouble with: `<REPLACE with e.g., Java quickstart>`
 - [x] If using Apache Kafka Java clients or a framework that uses Apache Kafka Java clients, version: Apache Spark Streaming 2.4.7 
 - [x] Kafka client configuration: `security.protocol=SASL_SSL, sasl.mechanism=PLAIN, sasl.jaas.config=XXX,startingOffsets=earliest,request.timeout.ms=60000,session.timeout.ms=30000`
 - [x] Namespace and EventHub/topic name: any
 - [x] Consumer or producer failure `kafka admin client api`
 - [ ] Timestamps *in UTC* `<REPLACE with e.g., Nov 7 2018 - 17:15:01 UTC>`
 - [x] group.id or client.id `autogenerated by spark on each run, something like this: spark-kafka-source-345180ff-bffd-4af5-a8a5-73ac36a4342c--906bb835-driver-0`
 - [x] Logs provided (with debug-level logging enabled if possible, e.g. log4j.rootLogger=DEBUG) or exception call stack
 - [x] Standalone repro `example of scripts and project to reproduce: https://github.com/aeftef/azure_eventhubs_kafka_issue`
 - [x] Operating system: `any, ubuntu 16.04, 18.04, windows 10`
 - [x] Critical issue

If this is a question on basic functionality, please verify the following:
 - [x] Port 9093 should not be blocked by firewall ("broker cannot be found" errors)
 - [x] Pinging FQDN should return cluster DNS resolution (e.g. `$ ping namespace.servicebus.windows.net` returns ~ `ns-eh2-prod-am3-516.cloudapp.net [13.69.64.0]`)
 - [x] Namespace should be either Standard or Dedicated tier, not Basic (TopicAuthorization errors):Standard  tier, auth and communication working

**Comment by arerlend** [comment]
@aeftef are you still seeing the issue?  I was investigating with the group 'spark-kafka-source-08133a25-9087-41e5-82e7-9782564a523e-5229197-driver-4' and was getting some strange behavior from the client.  I can't repro the NPE now with --describe --all-groups --all-topics anymore though.

**Comment by aeftef** [comment]
Hi Arthur,

Yes, we are still having this issue.

We can reproduce it at demand, for instance:

1. Creating a new event hub,
2. starting a spark streaming client(as the one in the example we provide)
3. And quering the consumer groups, for instance with the kafka cli utils (anything that uses the kafka admin to query the consumer groups api fails with a NPE in our scenario... I am thinking about kafdrop...).

The issue has also been reported via support ticket and AFAIK right now it should be at the product team support level, I can provide the support ticket id if it helps.

Thank you

**Comment by arerlend** [comment]
Yup, I'm investigating it now.  Let me try creating one on my own environment, thanks.

**Comment by sorbra** [comment]
Hi,

I have a potentially related issue. 

I am not using Spark streaming, but I do get a similar exception, so maybe my observations might help with debugging.

For context, I am using kafka-consumer-groups.sh to extract Kafka metrics from event hubs. Specifically, I need to know how many uncommitted messages exist for a given consumer group. (I use these metrics to control auto-scaling of Kafka consumer pods running in an Azure Kubernetes cluster). 

I run the following:

`kafka-consumer-groups.sh --describe --offsets --group "<consumer_group_name>" --bootstrap-server "<event_hub_namespace>.servicebus.windows.net:9093" --command-config "./config/sasl.properties"`

Now, when one or more consumer pods are running and listening on a given topic/consumer group, this works fine, and I get a nice list of offsets per partition for the topic the consumer group is subscribed to.

When the number of uncommitted messages drops to zero, Kubernetes auto-scaling kills off the last consumer for that consumer group, waiting for new messages to arrive. (In my application, this could be next week, which is why I am using auto-scaling to shutdown Kubernetes worker nodes during sleepy periods.).

After the last pod in a consumer group is scaled out of existence, the situation is this:
1. `kafka-consumer-groups.sh --list` still returns the name of the consumer group
2. `kafka-consumer-groups.sh --describe --offsets` fails with the same null pointer exception reported by aetf (slighly different line number, but could be caused by different kafka client versions):

```
java.lang.NullPointerException
        at kafka.admin.ConsumerGroupCommand$ConsumerGroupService.$anonfun$collectGroupsOffsets$6(ConsumerGroupCommand.scala:579)

```

**Comment by arerlend** [comment]
So @aeftef it looks like the issue is that the clients don't like us returning -1 offset for a consumer group that has members but no committed messages.  I'm looking into Kafka core code to see how they handle this case.

**Comment by arerlend** [comment]
@sorbra sounds like a slightly different scenario.  I can try to reproduce that as well.

**Comment by aeftef** [comment]
Hi Arthur @arerlend 

If event hubs is returning a -1 offset for a consumer group with no messages I agree and strongly suspect that is the cause of the problem.
Im forwarding to your mail an internal mail chain with azure support, we allready had, were we debugged kafka libraries to that point. I guess it can help.

**Comment by sorbra** [comment]
> @sorbra sounds like a slightly different scenario. I can try to reproduce that as well.

@arerlend : Thanks. On request from Microsoft Azure Support, I have created a [new issue on my scenario](https://github.com/Azure/azure-event-hubs-for-kafka/issues/143).

**Comment by arerlend** [comment]
@sorbra it's the same issue.

@aeftef you might have noticed that it started working for a bit, then stopped working.  We rolled out a change where we did not return topic partition offset if there were no committed offsets for given consumer group.  This broke a well known SDK so we had to roll back the change.  We will release an improved fix for the issue soon.

**Comment by arerlend** [comment]
@aeftef we have the fix.  When the admin client does the offset request, it's sending an empty list of topics.  In this case, we respond with all partitions, even those with -1 offsets, whereas Kafka only responds with the partitions with committed offsets.  Fix is in review and will be released soon.
